dbwrapper: properly destroy objects in case CDBWrapper ctor throws

Crypt-iQ opened the pull request:
There are two main changes: introduce LevelDbOptions to manage the leveldb::Options struct and changing CDBWrapper's penv member to a unique_ptr. The first allows LevelDbOptions to destroy leveldb::Options members without CDBWrapper having to directly. The second change allows the leveldb::Env to be managed by penv. Both of these changes prevent a LeakSanitizer exception if the CDBWrapper ctor throws.

Fixes https://github.com/bitcoin/bitcoin/issues/22592

mrnguyenthanhtu commented:
Have a nice day!

Crypt-iQ commented:
One build is failing due to a leak in `coins_view` and I cannot see why with the CI.  I am unable to reproduce this on my machine.

Talkless commented on src/dbwrapper.cpp:
```diff
@@ -97,46 +97,59 @@ static void SetMaxOpenFiles(leveldb::Options *options) {
-    leveldb::Status status = leveldb::DB::Open(options, path.string(), &pdb);
+    leveldb::Status status = leveldb::DB::Open(*options, path.string(), &pdb);
+    if (!status.ok()) {
+        delete penv;
```
If `penv` is deleted, then `~CDBWrapper()` gonna delete it again? Missing `= nullptr`?

Crypt-iQ replied:
If `!status.ok` then `HandleError` throws and the destructor isn't called

Talkless replied:
Maybe worth making `penv` a `unique_ptr` too, if we need manual handling with `delete`? Or do you think this would be kinda out of scope for this PR?

Adding addition `delete` into new code does not seem... "right" (https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rr-newdelete).

Crypt-iQ replied:
good point, I can look into making it a `unique_ptr`

Crypt-iQ replied:
The latest change makes `penv` a unique_ptr

ryanofsky commented on src/dbwrapper.cpp:
```diff
@@ -170,14 +183,7 @@ CDBWrapper::~CDBWrapper()
 {
     delete pdb;
```
In commit "dbwrapper: wrap options in unique_ptr and use DeleteOptions" (6770f066b69c6c271e7096cc647d5d42762f6b05)

I liked the suggestion to make `env` a unique_ptr https://github.com/bitcoin/bitcoin/pull/22663#discussion_r686783500, and maybe `pdb` could be a unique_ptr too?

Crypt-iQ replied:
I was not sure how to make `pdb` a unique_ptr and also pass it in to `Open` as `leveldb::DB**` when it is `nullptr` at this point. Maybe I am missing something.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/22663#discussion_r772488258

> I was not sure how to make `pdb` a unique_ptr and also pass it in to `Open` as `leveldb::DB**` when it is `nullptr` at this point. Maybe I am missing something.

No need to do this, but you would need to pass a temporary pointer to Open, and switch to unique_ptr after it returns:

```
leveldb::DB* new_db = nullptr;
leveldb::Status status = leveldb::DB::Open(options, fs::PathToString(path), &new_db);
pdb.reset(new_db);
```

ryanofsky commented on src/dbwrapper.cpp:
```diff
@@ -97,46 +97,59 @@ static void SetMaxOpenFiles(leveldb::Options *options) {
+}
+
+// Custom deleter for unique_ptr<leveldb::Options>.
+static void DeleteOptions(leveldb::Options* options)
```
In commit "dbwrapper: wrap options in unique_ptr and use DeleteOptions" (6770f066b69c6c271e7096cc647d5d42762f6b05)

Custom deleter seems to be forgetting to `delete options`, which would explain the fuzzing leak? https://github.com/bitcoin/bitcoin/pull/22663#issuecomment-894792761

Crypt-iQ replied:
No longer using a custom deleter and instead using `~LevelDbOptions`

ryanofsky reviewed (approved):
6770f066b69c6c271e7096cc647d5d42762f6b05 looks pretty good if the leak is fixed. It generally seems to moves things in a good direction.

I do think more ideally, if the goal is to make the options struct more RAII-like and exception-safe, instead of requiring it to be allocated separately on the heap and requiring it to be paired with a custom deleter, it would be better to just inherit from it and write a normal destructor:

```c++
struct LevelDbOptions : public leveldb::Options {
  ~LevelDbOptions() {
    delete filter_policy;
    delete info_log;
    ...
  }
}
```

laanwj commented on src/dbwrapper.cpp:
```diff
@@ -97,46 +97,59 @@ static void SetMaxOpenFiles(leveldb::Options *options) {
-    options.compression = leveldb::kNoCompression;
-    options.info_log = new CBitcoinLevelDBLogger();
+    options->block_cache = leveldb::NewLRUCache(nCacheSize / 2);
+    options->write_buffer_size = nCacheSize / 4; // up to two write buffers may be held in memory simultaneously
+    options->filter_policy = leveldb::NewBloomFilterPolicy(10);
+    options->compression = leveldb::kNoCompression;
+    options->info_log = new CBitcoinLevelDBLogger();
     if (leveldb::kMajorVersion > 1 || (leveldb::kMajorVersion == 1 && leveldb::kMinorVersion >= 16)) {
         // LevelDB versions before 1.16 consider short writes to be corruption. Only trigger error
         // on corruption in later versions.
-        options.paranoid_checks = true;
+        options->paranoid_checks = true;
     }
-    SetMaxOpenFiles(&options);
-    return options;
+    SetMaxOpenFiles(options);
+}
```
I agree with @ryanofsky. I don't think there's a strong need to change this function, but if there is, I would prefer it to return the target type `std::unique_ptr<leveldb::Options, void (*)(leveldb::Options*)` directly, instead of using a pointer out argument (which, strictly, seems a deterioration of API clarity to me).

Crypt-iQ replied:
Will work on a patch with this, thanks for the review

Crypt-iQ replied:
Latest change lifts the `GetOptions` logic into the ctor of `LevelDbOptions`

laanwj commented:
@Crypt-iQ can you please respond to @ryanofsky's comments and rebase ?

Crypt-iQ commented:
Hi yes, I can get to it this weekend

Crypt-iQ commented on src/dbwrapper.cpp:
```diff
@@ -97,46 +97,59 @@ static void SetMaxOpenFiles(leveldb::Options *options) {
+    options->info_log = nullptr;
+    delete options->block_cache;
+    options->block_cache = nullptr;
+    options->env = nullptr;
```
I opted for the second approach and dropped the line

Crypt-iQ commented:
I've rebased and hopefully addressed all of the comments

Crypt-iQ commented on src/dbwrapper.cpp:
```diff
@@ -97,36 +97,44 @@ static void SetMaxOpenFiles(leveldb::Options *options) {
-    options.filter_policy = leveldb::NewBloomFilterPolicy(10);
-    options.compression = leveldb::kNoCompression;
-    options.info_log = new CBitcoinLevelDBLogger();
+LevelDbOptions::LevelDbOptions() {
```
fixed

laanwj commented on src/dbwrapper.h:
```diff
@@ -19,6 +19,17 @@
+    ~LevelDbOptions();
+
+    /** Sets block_cache and write_buffer_size since nCacheSize isn't known when
+     * LevelDbOptions ctor is called.
```
I don't think this is strictly true? Or am I missing something? `nCacheSize` is passed into `CDBWrapper::CDBWrapper`, couldn't you pass it in like
```
CDBWrapper::CDBWrapper(const fs::path& path, size_t nCacheSize, bool fMemory, bool fWipe, bool obfuscate)
    : m_name{fs::PathToString(path.stem())},
      options(nCacheSize)
⋮

struct LevelDbOptions : public leveldb::Options {
    LevelDbOptions(size_t nCacheSize);
```

laanwj replied:
Thanks!

ryanofsky reviewed (approved):
Code review ACK 2a5d230c0fd764004a8c83c022ba71d0ee95d2a3. This all looks good and is straightforward. Thanks for implementing all the suggested changes

MarcoFalke commented on src/dbwrapper.h:
```diff
@@ -178,10 +184,10 @@ class CDBWrapper
 private:
     //! custom environment this database is using (may be nullptr in case of default environment)
-    leveldb::Env* penv;
+    std::unique_ptr<leveldb::Env> penv;
```
nit: Maybe rename to m_env while touching all lines? Maybe split the penv change into a separate commit for easier review?

Crypt-iQ replied:
renamed and split off into separate commit

Crypt-iQ replied:
I didn't functionally change the code and now there is a TSAN failure

MarcoFalke replied:
I couldn't make any sense of it, so reset it, but it is still here for anyone to look at: https://cirrus-ci.com/task/5612886578954240?logs=ci#L4869

Crypt-iQ replied:
I think this occurs and is unrelated to this PR because:
```
main thread:
* AppInitMain
  * ChainstateManager::ActiveTip()
    * CChain::Tip() -- accesses `vChain` without a lock (read)
```

```
b-loadblk thread:
* ThreadImport
  * ActivateBestChain called (src/node/blockstorage.cpp:883)
    * ActivateBestChainStep
      * ConnectTip
        * SetTip
          * vChain resize (write)
```

Crypt-iQ committed:
dbwrapper: make CDBWrapper use new LevelDbOptions

This change allows LevelDbOptions to destroy leveldb::Options
members instead of doing so in CDBWrapper's destructor. This change
prevents a LeakSanitizer exception if the CDBWrapper constructor
throws.

Crypt-iQ committed:
dbwrapper: change CDBWrapper penv to unique_ptr, rename to m_env

This changes the naming of penv to m_env and also makes it a
unique_ptr. Like the previous commit, this prevents a LeakSanitizer
exception if the constructor throws since the unique_ptr will clean
itself up.

achow101 commented:
cc @jamesob

Crypt-iQ commented:
unfortunately don't have time to work on this, lmk if i should close it

fanquake commented:
> unfortunately don't have time to work on this, lmk if i should close it

Thanks. Closed and added up-for-grabs.
